lib/src/command/add.dart

```diff
@@ -349,37 +454,141 @@ class AddCommand extends PubCommand {
         GitDescription(
           url: parsed.toString(),
           containingDir: p.current,
-          ref: gitRef,
-          path: gitPath,
+          ref: argResults.gitRef,
+          path: argResults.gitPath,
         ),
       );
     } else if (path != null) {
       ref = PackageRef(
           packageName, PathDescription(p.absolute(path), p.isRelative(path)));
-    } else if (sdk != null) {
-      ref = cache.sdk.parseRef(packageName, sdk);
+    } else if (argResults.sdk != null) {
+      ref = cache.sdk.parseRef(packageName, argResults.sdk);
     } else {
       ref = PackageRef(
         packageName,
         HostedDescription(
           packageName,
-          hostUrl ?? cache.hosted.defaultUrl,
+          argResults.hostedUrl ?? cache.hosted.defaultUrl,
         ),
       );
     }
-    return _ParseResult(ref, constraint);
+    return _PartialParseResult(ref, constraint);
+  }
+
+  /// Parse [package] to return the corresponding [_ParseResult].
+  ///
+  /// [package] must be written in the format
+  /// `<package-name>[:descriptor>]`, where quotations should be used if
+  /// necessary.
+  ///
+  /// `descriptor` is what you would put in a pubspec.yaml in the dependencies
+  /// section.
+  ///
+  /// Assumes that none of '--git-url', '--git-ref', '--git-path', '--path' and
+  /// '--sdk' are present in [argResults].
+  ///
+  ///
+  /// Examples:
+  /// ```
+  /// retry
+  /// retry:2.0.0
+  /// dev:retry:^2.0.0
+  /// retry:'>=2.0.0'
+  /// retry:'>2.0.0 <3.0.1'
+  /// 'retry:>2.0.0 <3.0.1'
+  /// retry:any
+  /// 'retry:{"path":"../foo"}'
+  /// 'retry:{"git":{"url":"../foo","ref":"branchname"},"version":"^1.2.3"}'
+  /// 'retry:{"sdk":"flutter"}'
+  /// 'retry:{"hosted":"mypub.dev"}'
```

Comment:
@sigurdm Is the syntax `package:{descriptor}` or `package{descriptor}`?

These lines indicate it's `package:{descriptor}`, but the examples in the help text show `package{descriptor}` (no colon between package and descriptor)

Are both allowed?

Reply:
I believe the `:` is required from my local testing and the help message mentions "by giving a descriptor after a colon".

I also noticed the help message includes single quotes (`'`) around the `package:{descriptor}` syntax which seems to not work for me. Double quotes do work (`"`) though.

Reply:
I've opened an issue at https://github.com/dart-lang/pub/issues/3749

Reply:
Yeah - you need the colon - thanks for spotting this!